testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java

```diff
@@ -0,0 +1,222 @@
+
+/**
+ * For tests that need to run under a specific server stability level,
+ * the server setup tasks from the inner classes can be used to change the stability level of the server to the desired level.
+ * Once the test is done, the original stability level is restored.
+ *
+ * In order to not pollute the configuration with XML from a different stability level following the run of the test,
+ * it takes a snapshot of the server configuration in the setup() method, and
+ */
+public abstract class StabilityServerSetupSnapshotRestoreTasks implements ServerSetupTask {
+
+    private final Stability desiredStability;
+    private volatile Stability originalStability;
+
+    private AutoCloseable snapshot;
+
+
+    public StabilityServerSetupSnapshotRestoreTasks(Stability desiredStability) {
+        this.desiredStability = desiredStability;
+    }
+
+    @Override
+    public final void setup(ManagementClient managementClient, String containerId) throws Exception {
+        // Make sure the desired stability level is one of the ones supported by the server
+        Set<Stability> supportedStabilityLevels = getSupportedStabilityLevels(managementClient);
+        Assume.assumeTrue(
+                String.format("%s is not a supported stability level", desiredStability, supportedStabilityLevels),
+                supportedStabilityLevels.contains(desiredStability));
+
+        // Check the reload-enhanced operation exists in the current stability level
+        Assume.assumeTrue(
+                "The reload-enhanced operation is not registered at this stability level",
+                checkReloadEnhancedOperationIsAvailable(managementClient));
+
+        // Check the reload-enhanced operation exists in the stability level we want to load to so that
+        // we can reload back to the current one
+        Stability reloadOpStability = getReloadEnhancedOperationStabilityLevel(managementClient);
+        Assume.assumeTrue(desiredStability.enables(reloadOpStability));
+
+
+        originalStability = readCurrentStability(managementClient.getControllerClient());
+
+        // We only want to reload to lower stability levels (e.g. when running the ts.preview tests, that contains
+        // some configuration at 'preview' level, so the reload to 'community' fails
+        Assume.assumeTrue(desiredStability.enables(originalStability));
+
+        // Take a snapshot, indicating that when reloading we want to go back to the original stability
+        snapshot = takeSnapshot(managementClient, originalStability);
+
+        // All good, let's do it!
+        reloadToDesiredStability(managementClient.getControllerClient(), desiredStability);
+
+        // Do any additional setup from the sub-classes
+        doSetup(managementClient);
+    }
+
+    protected void doSetup(ManagementClient managementClient) throws Exception {
+
+    }
+
+    @Override
+    public void tearDown(ManagementClient managementClient, String containerId) throws Exception {
```

Comment:
Should this be a final method with a `doTearDown()` to ensure the snapshot is always restored?